src/tag.cc

```diff
@@ -51,152 +46,185 @@ Tag::~Tag() {
+NAN_METHOD(Tag::IsEmpty) {
+    Tag *t = ObjectWrap::Unwrap<Tag>(info.Holder());
+    info.GetReturnValue().Set(t->tag->isEmpty());
 }
 
-void Tag::SetGenre(Local<String> property, Local<Value> value, const AccessorInfo& info) {
-  HandleScope scope;
-  unwrapTag(info)->tag->setGenre(NodeStringToTagLibString(value));
+NAN_METHOD(Tag::SyncSaveTag) {
+    Tag *t = ObjectWrap::Unwrap<Tag>(info.Holder());
+    assert(t->fileRef);
+    bool success = t->fileRef->save();
+    if (success)
+        info.GetReturnValue().SetUndefined();
+    else
+        Nan::ThrowError("Failed to save file.");
+        //TODO: filename
 }
 
-Handle<Value> Tag::IsEmpty(const Arguments &args) {
-  HandleScope scope;
-  Tag *t = ObjectWrap::Unwrap<Tag>(args.This());
-  return Boolean::New(t->tag->isEmpty());
+NAN_METHOD(Tag::AsyncSaveTag) {
+    if (info.Length() >= 1 && !info[0]->IsFunction()) {
+        Nan::ThrowError("Expected callback function as first argument");
+        return;
+    }
+
+    Local<Function> callback = Local<Function>::Cast(info[0]);
+
+    Tag *t = ObjectWrap::Unwrap<Tag>(info.Holder());
+
+    AsyncBaton *baton = new AsyncBaton;
+    baton->request.data = baton;
+    baton->tag = t;
+    baton->callback.Reset(callback);
+    baton->error = 1;
+
+    uv_queue_work(uv_default_loop(), &baton->request, Tag::AsyncSaveTagDo, (uv_after_work_cb)Tag::AsyncSaveTagAfter);
+
+    info.GetReturnValue().SetUndefined();
 }
 
-Handle<Value> Tag::SyncSaveTag(const Arguments &args) {
-  HandleScope scope;
-  Tag *t = ObjectWrap::Unwrap<Tag>(args.This());
-  assert(t->fileRef);
-  bool success = t->fileRef->save();
-  if (success)
-      return Undefined();
-  else
-      return ThrowException(String::Concat(
-                String::New("Failed to save file: "),
-                String::New(t->fileRef->file()->name())
-            ));
+void Tag::AsyncSaveTagDo(uv_work_t *req) {
+    AsyncBaton *baton = static_cast<AsyncBaton*>(req->data);
+
+    assert(baton->tag->fileRef);
+    baton->error = !baton->tag->fileRef->save();
 }
 
-Handle<Value> Tag::SyncTag(const Arguments &args) {
-    HandleScope scope;
+void Tag::AsyncSaveTagAfter(uv_work_t *req) {
+    Isolate* isolate = Isolate::GetCurrent();
+    HandleScope scope(isolate);
+
+    AsyncBaton *baton = static_cast<AsyncBaton*>(req->data);
+
+    if (baton->error) {
+        Local<Object> error = Nan::New<Object>();
+        error->Set(Nan::New("message").ToLocalChecked(), Nan::New("Failed to save file").ToLocalChecked());
+        error->Set(Nan::New("path").ToLocalChecked(), Nan::New(baton->tag->fileRef->file()->name()).ToLocalChecked());
+        Handle<Value> argv[] = { error };
+        Nan::Call(Nan::New(baton->callback), Nan::GetCurrentContext()->Global(), 1, argv);
+    }
+    else {
+        Handle<Value> argv[] = { Nan::Null() };
+        Nan::Call(Nan::New(baton->callback), Nan::GetCurrentContext()->Global(), 1, argv);
+    }
 
+    //baton->callback.Dispose();
```

Comment:
How does the reference to the callback get released when using NaN without this dispose call?